### PR TITLE
Fix bug when finest_level was less then max_radial_level. 

### PR DESCRIPTION
**Pull request by ChrisDeGrendele**
This PR addresses #203. There are two main things it changes - 
1. The AMReX MLMG solvers were using the size of the geometry object (which is initialized to max_radial_level) to loop over grids. Now we create small_geoms / small_grids which only are defined up to finest_level. This prevents the linear solvers from looping over undefined grids (In the case that we haven't actually used all the levels yet. 
2. We now assume that at the start, finest_radial_level is 0, rather than max_radial_level. It then gets updated upon regridding, but will always initialize data on the coarsest grid. 

- I also added a print box array function in maestrodebug.cpp, because I used this to helped me debug this issue. 

Test Results:
(I used AMReX/Tools/Plotfiles/fcomare.cpp)
- I found that mach_jet agrees well between amr_levels on this branch and from the previous version of mach_jet. 
- One potetional problem: `tpert` seems to grow the most error, at almost 15%. 
Rerunning these tests, with `maestro.use_tpert_in_tagging` I got roughly the same errors as below between lvl0 and lvl2, I'm not sure if this a cause for concern?

The tests:
- Running the mach_jet file setup as is, with `inputs_2d` and amr.max_level = 0,1,2 
   - `mach_jet_InitData` plotfile agreed with 0 error between all 3 levels. 

- For the plotfile `mach_jet0000000` comparing amr.max_level = 0,1,2
   - LVL1 and LVL2 agreed completley. 

  - LVL 0 vs LVL 1 `mach_jet0000000`
```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 velx                               8.082423619e-14           4.279259831e-17
 vely                               2.273736754e-13            6.54637773e-17
 magvel                             4.547473509e-13           1.288152866e-16
 momentum                           4.440892099e-16           1.257961784e-16
 vort                                3.91153776e-12           2.668249665e-17
 rho                                              0                         0
 rhoh                                             0                         0
 h                                                0                         0
 rhoX(A)                                          0                         0
 rhoX(B)                                          0                         0
 X(A)                                             0                         0
 X(B)                                             0                         0
 abar                                             0                         0
 omegadot(A)                                      0                         0
 omegadot(B)                                      0                         0
 Hnuc                                             0                         0
 tfromp                                           0                         0
 tfromh                                           0                         0
 deltap                                           0                         0
 deltaT                                           0                         0
 Pi                                 3.793466796e-08           1.067608332e-11
 pioverp0                           3.793466796e-14           1.067608332e-11
 p0pluspi                           3.795139492e-08            3.79489157e-14
 gpix                               9.406404544e-10           2.189717414e-14
 gpiy                               6.074217413e-07           8.483977717e-12
 rhopert                                          0                         0
 rhohpert                                         0                         0
 tpert                                            0                         0
 rho0                                             0                         0
 rhoh0                                            0                         0
 h0                                               0                         0
 p0                                               0                         0
 MachNumber                         1.040834086e-17           1.203658002e-16
 deltagamma                                       0                         0
 entropy                                          0                         0
 entropypert                                      0                         0
 S                                                0                         0
 w0x                                              0                         0
 w0y                                              0                         0
 divw0                                            0                         0
 thermal                                          0                         0
 conductivity                                     0                         0
```

For LVL 0 vs LVL2 
```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 velx                               8.082423619e-14           4.279259831e-17
 vely                               2.273736754e-13            6.54637773e-17
 magvel                             4.547473509e-13           1.288152866e-16
 momentum                           4.440892099e-16           1.257961784e-16
 vort                                3.91153776e-12           2.668249665e-17
 rho                                              0                         0
 rhoh                                             0                         0
 h                                                0                         0
 rhoX(A)                                          0                         0
 rhoX(B)                                          0                         0
 X(A)                                             0                         0
 X(B)                                             0                         0
 abar                                             0                         0
 omegadot(A)                                      0                         0
 omegadot(B)                                      0                         0
 Hnuc                                             0                         0
 tfromp                                           0                         0
 tfromh                                           0                         0
 deltap                                           0                         0
 deltaT                                           0                         0
 Pi                                 3.793466796e-08           1.067608332e-11
 pioverp0                           3.793466796e-14           1.067608332e-11
 p0pluspi                           3.795139492e-08            3.79489157e-14
 gpix                               9.406404544e-10           2.189717414e-14
 gpiy                               6.074217413e-07           8.483977717e-12
 rhopert                                          0                         0
 rhohpert                                         0                         0
 tpert                                            0                         0
 rho0                                             0                         0
 rhoh0                                            0                         0
 h0                                               0                         0
 p0                                               0                         0
 MachNumber                         1.040834086e-17           1.203658002e-16
 deltagamma                                       0                         0
 entropy                                          0                         0
 entropypert                                      0                         0
 S                                                0                         0
 w0x                                              0                         0
 w0y                                              0                         0
 divw0                                            0                         0
 thermal                                          0                         0
```
- The last test is at timstep 55, plot file `mach_jet0000055` comparing amr.max_level=0,1,2
  - Again LVL1 and LVL2 completley agree
   - LVL0 vs LVL 1 
```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 velx                               4.831690603e-12           3.328007366e-15
 vely                               5.456968211e-12           1.161818115e-15
 magvel                             4.661160347e-12           9.923795546e-16
 momentum                           1.032507413e-14           2.198249299e-15
 vort                               9.458744898e-11           2.624366366e-15
 rho                                1.452830911e-17           1.452830911e-14
 rhoh                               9.313225746e-09           3.725290298e-15
 h                                  3.528594971e-05           1.411437988e-14
 rhoX(A)                            1.452830911e-17           1.452830911e-14
 rhoX(B)                                          0                         0
 X(A)                                             0                         0
 X(B)                                             0                         0
 abar                                             0                         0
 omegadot(A)                                      0                         0
 omegadot(B)                                      0                         0
 Hnuc                                             0                         0
 tfromp                             1.758593271e-13           1.462175718e-14
 tfromh                             1.705302566e-13           1.417867363e-14
 deltap                             3.725290298e-09           0.0006647554946
 deltaT                             3.544668408e-15           0.0006325110689
 Pi                                 3.092281986e-11            5.98520144e-15
 pioverp0                           3.035766083e-17           5.875813271e-15
 p0pluspi                           1.164153218e-10           1.163929382e-16
 gpix                               1.273292582e-10           5.599417365e-15
 gpiy                               1.746229827e-10           6.126854525e-15
 rhopert                            1.452830911e-17            0.002590573406
 rhohpert                           9.313225746e-09           0.0006647389238
 tpert                              1.598721155e-13              0.1490066225
 rho0                                             0                         0
 rhoh0                                            0                         0
 h0                                               0                         0
 p0                                               0                         0
 MachNumber                         1.387778781e-16           1.206225159e-15
 deltagamma                                       0                         0
 entropy                            3.039836884e-06           7.216581262e-15
 entropypert                        6.650574888e-15            0.006606691032
 S                                                0                         0
 w0x                                              0                         0
 w0y                                              0                         0
 divw0                                            0                         0
 thermal                                          0                         0
 conductivity                                     0                         0
```
And LVL 0 vs LVL 2
```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 velx                               4.831690603e-12           3.328007366e-15
 vely                               5.456968211e-12           1.161818115e-15
 magvel                             4.661160347e-12           9.923795546e-16
 momentum                           1.032507413e-14           2.198249299e-15
 vort                               9.458744898e-11           2.624366366e-15
 rho                                1.452830911e-17           1.452830911e-14
 rhoh                               9.313225746e-09           3.725290298e-15
 h                                  3.528594971e-05           1.411437988e-14
 rhoX(A)                            1.452830911e-17           1.452830911e-14
 rhoX(B)                                          0                         0
 X(A)                                             0                         0
 X(B)                                             0                         0
 abar                                             0                         0
 omegadot(A)                                      0                         0
 omegadot(B)                                      0                         0
 Hnuc                                             0                         0
 tfromp                             1.758593271e-13           1.462175718e-14
 tfromh                             1.705302566e-13           1.417867363e-14
 deltap                             3.725290298e-09           0.0006647554946
 deltaT                             3.544668408e-15           0.0006325110689
 Pi                                 3.092281986e-11            5.98520144e-15
 pioverp0                           3.035766083e-17           5.875813271e-15
 p0pluspi                           1.164153218e-10           1.163929382e-16
 gpix                               1.273292582e-10           5.599417365e-15
 gpiy                               1.746229827e-10           6.126854525e-15
 rhopert                            1.452830911e-17            0.002590573406
 rhohpert                           9.313225746e-09           0.0006647389238
 tpert                              1.598721155e-13              0.1490066225
 rho0                                             0                         0
 rhoh0                                            0                         0
 h0                                               0                         0
 p0                                               0                         0
 MachNumber                         1.387778781e-16           1.206225159e-15
 deltagamma                                       0                         0
 entropy                            3.039836884e-06           7.216581262e-15
 entropypert                        6.650574888e-15            0.006606691032
 S                                                0                         0
 w0x                                              0                         0
 w0y                                              0                         0
 divw0                                            0                         0
 thermal                                          0                         0
 conductivity                                     0                         0
```

- One last test I did is to compare the current main branch of MAESTROeX of mach_jet with amr.max_level = 0 (it didn't run with higher levels), with amr.max_level=2 of this PR.

```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 velx                               4.831690603e-12           3.328007366e-15
 vely                               5.456968211e-12           1.161818115e-15
 magvel                             4.661160347e-12           9.923795546e-16
 momentum                           1.032507413e-14           2.198249299e-15
 vort                               9.458744898e-11           2.624366366e-15
 rho                                1.452830911e-17           1.452830911e-14
 rhoh                               9.313225746e-09           3.725290298e-15
 h                                  3.528594971e-05           1.411437988e-14
 rhoX(A)                            1.452830911e-17           1.452830911e-14
 rhoX(B)                                          0                         0
 X(A)                                             0                         0
 X(B)                                             0                         0
 abar                                             0                         0
 omegadot(A)                                      0                         0
 omegadot(B)                                      0                         0
 Hnuc                                             0                         0
 tfromp                             1.758593271e-13           1.462175718e-14
 tfromh                             1.705302566e-13           1.417867363e-14
 deltap                             3.725290298e-09           0.0006647554946
 deltaT                             3.544668408e-15           0.0006325110689
 Pi                                 3.092281986e-11            5.98520144e-15
 pioverp0                           3.035766083e-17           5.875813271e-15
 p0pluspi                           1.164153218e-10           1.163929382e-16
 gpix                               1.273292582e-10           5.599417365e-15
 gpiy                               1.746229827e-10           6.126854525e-15
 rhopert                            1.452830911e-17            0.002590573406
 rhohpert                           9.313225746e-09           0.0006647389238
 tpert                              1.598721155e-13              0.1507537688
 rho0                                             0                         0
 rhoh0                                            0                         0
 h0                                               0                         0
 p0                                               0                         0
 MachNumber                         1.387778781e-16           1.206225159e-15
 deltagamma                                       0                         0
 entropy                            3.039836884e-06           7.216581262e-15
 entropypert                        6.650574888e-15            0.006604834177
 S                                                0                         0
 w0x                                              0                         0
 w0y                                              0                         0
 divw0                                            0                         0
 thermal                                          0                         0
 conductivity                                     0                         0
```

### CODE REVIEW

File: Source/MaestroThermal.cpp [hunk 1]
```diff
@@ -159,7 +172,20 @@ void Maestro::MakeExplicitThermalHterm(Vector<MultiFab>& thermal,
     // turn off multigrid coarsening since no actual solve is performed
     info.setMaxCoarseningLevel(0);
 
-    MLABecLaplacian mlabec(geom, grids, dmap, info);
+    amrex::Vector<amrex::Geometry> small_geom(finest_level+1);
```
Comment (dwillcox) [inline Source/MaestroThermal.cpp:175]: we can do this to construct the small vectors, e.g.:

```
amrex::Vector<amrex::Geometry> small_geom(geom.begin(), geom.begin() + finest_level + 1);
```

Comment (WeiqunZhang) [reply Source/MaestroThermal.cpp:175]: Maestro inherits a function from AmrCore, `Vector<Geometry> Geom (int a_coarsest_lev, int a_finest_lev) const`.

**Comment by ajnonaka** [comment]
@ChrisDeGrendele note that finest_radial_level exists in both the fortran namelist and the C++ namespace (separate copies; they don't "point" to each other).  In BaseStateGeometry.cpp you set it to 0, whereas in base_state_geometry.F90 it is set to max_radial_level.  I'm wondering if this is going to be a problem.

I also wonder if at the beginning of regridding, we always set finest_radial_level to 0 (and update it) and make sure the fortran namelist version gets updated accordingly as well.

**Comment by ajnonaka** [comment]
@ChrisDeGrendele The tpert error being 15% is not a problem.  Look at the absolute difference; it is 1.598721155e-13.  This likely means that in a particular cell, tpert was very close to zero, so a significant bit can be 15% the size of the signal.  This is totally ok.

File: Source/MaestroThermal.cpp (hunk 1 above)
Comment (dwillcox) [reply Source/MaestroThermal.cpp:175]: ah, that's easy, thank you!

**Comment by ChrisDeGrendele** [comment]
In Maestro::Regrid, the `base_geom.finest_radial_level` is set equal to `finest_level`.

When we call `init_multilevel(tag_array.dataPtr(), &finest_level)`, this is bound to the fortran function, in base_state_geometry.F90. This takes, the c++ finest_level and sets `finest_radial_level` in fortran. 

So, three things need to be made sure.
1. We start with the same assumptions in c++ / fortran (change fortran to start with lvl 0 as well)
2. Whenever C++ updates the `base_geom.finest_radial_level` in Regrid(), `init_multilevel(tag_array.dataPtr(), &finest_level)` needs to be called as well. This is already implemented. 
3. If fortran updates, finest_radial_level, we need to send this back to c++. I don't think it does ever, there is no regridding implemented in fortran. 

Correct me if I'm wrong, but I think the fortran finest_radial_level should always be synced now with c++.

File: Source/MaestroDebug.cpp [hunk 2]
```diff
@@ -81,6 +83,31 @@ void Maestro::PrintMF(const Vector<MultiFab>& MF) {
...
+      for (int i = 0; i < ba[lev].size(); ++i) {
+	  Print() << "Grid #" << i << std::endl;
+	
+	  const Box& validBox = ba[lev][i];
+	  auto lo = validBox.loVect3d();
+	  auto hi = validBox.hiVect3d();
+
+	  std::cout << "Level " << lev << std::endl;
+	  std::cout << "valid box ";
+	  for (auto n = 0; n < AMREX_SPACEDIM; ++n) {
```
Comment (ajnonaka) [inline Source/MaestroDebug.cpp:102]: Be careful mixing Print() and std::cout throughout this routine.
Maybe this whole routine should be in an IOProcessor if test and then just use cout?
Each processor has a copy of the boxarray.

File: Source/MaestroInit.cpp [hunk 3]
```diff
@@ -462,7 +460,7 @@ void Maestro::InitProj() {
     Make_S_cc(S_cc_old, delta_gamma1_term, delta_gamma1, sold, uold,
               rho_omegadot, rho_Hnuc, rho_Hext, thermal, p0_old, gamma1bar_old,
               delta_gamma1_termbar);
-
+
```
Comment (ajnonaka) [inline Source/MaestroInit.cpp:463]: Can you revert the changes to this file (doing by hand is fine)?  The changes were deleting new lines and adding whitespace, neither of which were needed.

File: Source/MaestroMacProj.cpp [hunk 4]
```diff
@@ -132,11 +132,14 @@ void Maestro::MacProj(Vector<std::array<MultiFab, AMREX_SPACEDIM> >& umac,
         }
     }
 
-    //
+
+
```
Comment (ajnonaka) [inline Source/MaestroMacProj.cpp:136]: Clean up this extra white space, thanks.

File: Source/MaestroThermal.cpp [hunk 5]
```diff
@@ -46,7 +46,20 @@ void Maestro::MakeExplicitThermal(
...
+    amrex::Vector<amrex::BoxArray> small_grids(finest_level+1);
+    amrex::Vector<amrex::DistributionMapping> small_dmap(finest_level+1);
+    amrex::Vector<amrex::MultiFab> small_rhcc(finest_level+1);
```
Comment (ajnonaka) [inline Source/MaestroThermal.cpp:52]: small_rhcc doesn't appear to be used.  delete all 4 instances in MaestroThermal.cpp

**Comment by zingale** [comment]
this changes one problem in our suite: 
http://groot.astro.sunysb.edu/MAESTROeX/test-suite/gfortran/2020-07-28-001/reacting_bubble-2d-amr.html
